src/mgr/Action.h

```diff
@@ -76,11 +78,16 @@ class Action: public RefCountable
     /// calculate and keep local action-specific information
     virtual void collect() {}
 
+    /// write manager report output to a stream.
+    virtual void print(std::ostream &) {}
```

Comment:
I am expecting the formater will be something(s) presented by the Action which these display methods can make use of - either as object to write to the stream as-needed, or methods/functions to prettify specific datums before they are added to the stream.
None of the options on implementing that properly require this method to know more than its parameter is an ostream based type.

Reply:
> I am expecting the formater will be something(s) presented by the Action which these display methods can make use of

With that design, print() should take no parameters at all! The method should just access a ReportStream data member. I have doubts about that design because it forces that data member to exist outside of print() call/stage, probably wrapped in std::optional member (that we will then need to dereference and keep in sync with StoreEntry). A print() parameter expresses the printing context/scope better AFAICT.

>  None of the options on implementing that properly require this method to know more than its parameter is an ostream based type.

We seem to agree that printers will need additional state. That additional state should be stored somewhere. I see no advantages (and several serious disadvantages!) in storing part of the printing state (i.e. the stream) in a parameter and part (e.g., indentation level) in a class data member.